### PR TITLE
[FLINK-4692] [tableApi] Add tumbling group-windows for batch tables

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 1]
```diff
@@ -84,14 +86,153 @@ object AggregateUtil {
...
+    groupings: Array[Int],
+    inputType: RelDataType): MapFunction[Any, Row] = {
+
+    val (aggFieldIndexes, aggregates) = transformToAggregateFunctions(
+      namedAggregates.map(_.getKey),
+      inputType,
+      groupings.length)
+
+    val mapReturnType: RowTypeInfo =
+      createAggregateBufferDataType(groupings, aggregates, inputType, Some(Types.LONG))
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:125]: BasicTypeInfo.LONG_TYPE_INFO

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala [hunk 2]
```diff
@@ -55,7 +55,21 @@ abstract class EventTimeGroupWindow(
   }
 }
 
-abstract class ProcessingTimeGroupWindow(name: Option[Expression]) extends LogicalWindow(name)
+abstract class ProcessingTimeGroupWindow(name: Option[Expression]) extends LogicalWindow(name) {
+  override def validate(tableEnv: TableEnvironment): ValidationResult = {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:59]: 是否可以全部之后检查，统一到transplan？

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala (hunk 1 above)
Comment (wuchong) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala:125]: 我觉得`Types.LONG`更好， Types 是 table api 里封装的util。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    val config = tableEnv.getConfig
+
+    val inputDS = getInput.asInstanceOf[DataSetRel].translateToPlan(
+      tableEnv,
+      // tell the input operator that this operator currently only supports Rows as input
+      Some(TypeConverter.DEFAULT_ROW_TYPE))
+
+    val result = window match {
+      case EventTimeTumblingGroupWindow(_, _, size) =>
+        createEventTimeTumblingWindowDataSet(inputDS, isTimeInterval(size.resultType))
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala:123]: 很棒。需要讨论一下。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      } else {
+        // TODO: count tumbling all window on event-time should sort all the data set
+        // on event time before applying the windowing logic.
+        throw new UnsupportedOperationException(
+          "Count tumbling non-grouping window on event-time are currently not supported.")
+      }
+    }
+  }
+
+  private def prepareOperatorName: String = {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala:210]: 下面是很棒的 抽取。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala [hunk 5]
```diff
@@ -512,24 +653,22 @@ object AggregateUtil {
   }
 
   private def createAggregateBufferDataType(
-    groupings: Array[Int],
-    aggregates: Array[Aggregate[_]],
-    inputType: RelDataType): RowTypeInfo = {
+      groupings: Array[Int],
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/AggregateUtil.scala]: 缩进

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+/** This map function only works for windows on batch tables. The differences between this function
+  * and [[org.apache.flink.api.table.runtime.aggregate.AggregateMapFunction]] is this function
+  * append a rowtime field to the end of the output row.
+  */
+class DataSetWindowAggregateMapFunction(
+    private val aggregates: Array[Aggregate[_]],
+    private val aggFields: Array[Int],
+    private val groupingKeys: Array[Int],
+    private val timeFieldPos: Int,    // time field position in input row
+    private val tumbleTimeWindowSize: Option[Long],
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala:41]: 这块应该通用一下。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    val aggString = aggregationToString(inputType, grouping, getRowType, namedAggregates, Nil)
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala]: Nil？namedProperties

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.flink.api.table.Row
+import org.apache.flink.api.table.runtime.aggregate.Aggregate
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow
+import org.apache.flink.util.Preconditions
+
+
+/** This map function only works for windows on batch tables. The differences between this function
+  * and [[org.apache.flink.api.table.runtime.aggregate.AggregateMapFunction]] is this function
+  * append a rowtime field to the end of the output row.
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala:34]: row-time

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    } else {
+      // otherwise, set rowtime for future use
+      output.setField(rowtimeIndex, rowtime)
+    }
+
+    output
+  }
+
+  private def getTimestamp(timeField: Any): Long = {
+    timeField match {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala]: 感觉FlinkTypeFactory#typeInfoToSqlTypeName 对类型处理更严谨一些。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala (hunk 2 above)
Comment (wuchong) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/groupWindows.scala:59]: 你说在 translateToPlan 中检查？ 我觉得不好，validate 阶段要做的是 语法检测， translateToPlan 中抛的异常主要是暂不支持。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala (hunk 7 above)
Comment (wuchong) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetWindowAggregate.scala]: 哦， 漏了 namedProperties

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala (hunk 9 above)
Comment (wuchong) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/runtime/aggregate/window/batch/DataSetWindowAggregateMapFunction.scala:86]: 这块讨论一下

**Pull request closed without merging**